daemon/containerd/cache.go

```diff
@@ -3,10 +3,103 @@ package containerd
 import (
 	"context"
 
+	"github.com/docker/docker/api/types/container"
+	imagetype "github.com/docker/docker/api/types/image"
 	"github.com/docker/docker/builder"
+	"github.com/docker/docker/image"
 )
 
 // MakeImageCache creates a stateful image cache.
 func (i *ImageService) MakeImageCache(ctx context.Context, cacheFrom []string) (builder.ImageCache, error) {
-	panic("not implemented")
+	images := []*image.Image{}
+	for _, c := range cacheFrom {
+		im, err := i.GetImage(context.TODO(), c, imagetype.GetImageOpts{})
+		if err != nil {
+			return nil, err
+		}
+		images = append(images, im)
+	}
+	return &imageCache{images: images, c: i}, nil
+}
+
+type imageCache struct {
+	images []*image.Image
+	c      *ImageService
+}
+
+// todo: this is never returning a hit now :')
+func (ic *imageCache) GetCache(parentID string, cfg *container.Config) (imageID string, err error) {
+	i, err := ic.c.GetImage(context.TODO(), parentID, imagetype.GetImageOpts{})
+	if err != nil {
+		for _, ii := range ic.images {
+			if ii.ID().String() == parentID {
+				if compare(ii.RunConfig(), cfg) {
+					return ii.ID().String(), nil
+				}
+			}
+		}
+	} else {
+		if compare(i.RunConfig(), cfg) {
+			return i.ID().String(), nil
+		}
+	}
+	return "", nil
+}
+
+// compare two Config struct. Do not compare the "Image" nor "Hostname" fields
+// If OpenStdin is set, then it differs
+func compare(a, b *container.Config) bool {
+	if a == nil || b == nil ||
+		a.OpenStdin || b.OpenStdin {
+		return false
+	}
+	if a.AttachStdout != b.AttachStdout ||
+		a.AttachStderr != b.AttachStderr ||
+		a.User != b.User ||
+		a.OpenStdin != b.OpenStdin ||
+		a.Tty != b.Tty {
+		return false
+	}
+
+	if len(a.Cmd) != len(b.Cmd) ||
+		len(a.Env) != len(b.Env) ||
+		len(a.Labels) != len(b.Labels) ||
+		len(a.ExposedPorts) != len(b.ExposedPorts) ||
+		len(a.Entrypoint) != len(b.Entrypoint) ||
+		len(a.Volumes) != len(b.Volumes) {
+		return false
+	}
+
+	for i := 0; i < len(a.Cmd); i++ {
+		if a.Cmd[i] != b.Cmd[i] {
+			return false
+		}
+	}
+	for i := 0; i < len(a.Env); i++ {
+		if a.Env[i] != b.Env[i] {
+			return false
+		}
+	}
+	for k, v := range a.Labels {
+		if v != b.Labels[k] {
+			return false
+		}
+	}
+	for k := range a.ExposedPorts {
+		if _, exists := b.ExposedPorts[k]; !exists {
+			return false
+		}
+	}
+
+	for i := 0; i < len(a.Entrypoint); i++ {
+		if a.Entrypoint[i] != b.Entrypoint[i] {
+			return false
+		}
+	}
+	for key := range a.Volumes {
+		if _, exists := b.Volumes[key]; !exists {
+			return false
+		}
+	}
+	return true
```

Comment:
We already have google/go-cmp vendored so we could use `cmp.Equal` (https://pkg.go.dev/github.com/google/go-cmp/cmp#Equal).
https://go.dev/play/p/vptRlAFtS96

```suggestion
func compare(a, b *container.Config) bool {
	if a == nil || b == nil || a.OpenStdin || b.OpenStdin {
			return false
	}
    return cmp.Equal(a, b, cmpopts.IgnoreFields(container.Config{}, "Hostname", "Image"))
```

Reply:
Hmm, on the other hand:
> It is intended to only be used in tests, as performance is not a goal and it may panic if it cannot compare the values.

So maybe not the best idea.

Reply:
We could! That does look a lot better. I think it's worth calling out that `reflect.DeepEqual` will always perform worse than writing out all the checks manually, but for that size struct that probably doesn't matter much.

Reply:
Do you think it's worth doing a benchmark of that against something like marshalling both to JSON for comparison? :eyes:

(I bet `reflect.DeepEqual` is faster, but who knows :sweat_smile:)